File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int Enter(int num, int k, vector<vector<int>>& d, vector<bool> visit)
{
    int total = 0;
    visit[num] = true;
    k -= d[num][1];
    for(int i = 0; i < d.size(); i++)
    {
        if(i == num || visit[i]) continue;
        if(k < d[i][0]) continue;
        
        total = max(total, Enter(i, k, d, visit));
    }
    return total + 1;
}
int solution(int k, vector<vector<int>> dungeons) {
    int answer = 0;
    vector<bool> visit(dungeons.size(), false);
    
    for(int i = 0; i < dungeons.size(); i++)
    {
        if(k < dungeons[i][0]) continue;
        answer = max(answer, Enter(i, k, dungeons, visit));
    }
    
    return answer;
}

File: 프로그래머스/2/87946. 피로도/피로도.cs
using System;

public class Solution {
    public int solution(int k, int[,] dungeons) {
        bool[] visit = new bool[dungeons.GetLength(0)];
        return FindDungeon(k, dungeons, 0, visit);
    }
    public int FindDungeon(int k, int[,] dungeons, int count, bool[] visit)
    {
        if(count >= dungeons.GetLength(0)) return count;
        int answer = count;
        for(int i = 0; i < dungeons.GetLength(0); i++)
        {
            if(visit[i]) continue;
            if(k - dungeons[i,0] < 0) continue;
            bool[] newVisit = new bool[dungeons.GetLength(0)];
            Array.Copy(visit, newVisit, dungeons.GetLength(0));
            newVisit[i] = true;
            int temp = FindDungeon(k-dungeons[i,1], dungeons, count + 1, newVisit);
            answer = answer > temp ? answer : temp;
        }
        return answer;
        
    }
}